clang/lib/CIR/CodeGen/LoweringPrepare.cpp

```diff
@@ -865,30 +893,369 @@ void LoweringPreparePass::lowerThreeWayCmpOp(CmpThreeWayOp op) {
+    // since MLIR doesn't allow public declarations.
+    mlir::SymbolTable::setSymbolVisibility(
+        g, mlir::SymbolTable::Visibility::Private);
+  }
+  return g;
+}
+
+static cir::FuncOp getGuardAbortFn(clang::CIRGen::CIRGenModule &cgm,
+                                   cir::PointerType guardPtrTy) {
+  // void __cxa_guard_abort(__guard *guard_object);
+  cir::FuncType fTy = cgm.getBuilder().getFuncType(guardPtrTy, {cgm.VoidTy},
+                                                   /*isVarArg=*/false);
+  assert(!cir::MissingFeatures::functionIndexAttribute());
+  assert(!cir::MissingFeatures::noUnwindAttribute());
+  return cgm.createRuntimeFunction(fTy, "__cxa_guard_abort");
+}
+
+static cir::FuncOp getGuardAcquireFn(clang::CIRGen::CIRGenModule &cgm,
+                                     cir::PointerType guardPtrTy) {
+  // int __cxa_guard_acquire(__guard *guard_object);
+  cir::FuncType fTy = cgm.getBuilder().getFuncType(
+      guardPtrTy, {cgm.getTypes().ConvertType(cgm.getASTContext().IntTy)},
+      /*isVarArg=*/false);
+  assert(!cir::MissingFeatures::functionIndexAttribute());
+  assert(!cir::MissingFeatures::noUnwindAttribute());
+  return cgm.createRuntimeFunction(fTy, "__cxa_guard_acquire");
+}
+
+static cir::FuncOp getGuardReleaseFn(clang::CIRGen::CIRGenModule &cgm,
+                                     cir::PointerType guardPtrTy) {
+  // void __cxa_guard_release(__guard *guard_object);
+  cir::FuncType fTy = cgm.getBuilder().getFuncType(guardPtrTy, {cgm.VoidTy},
+                                                   /*isVarArg=*/false);
+  assert(!cir::MissingFeatures::functionIndexAttribute());
+  assert(!cir::MissingFeatures::noUnwindAttribute());
+  return cgm.createRuntimeFunction(fTy, "__cxa_guard_release");
+}
+
+namespace {
+struct CallGuardAbort final : clang::CIRGen::EHScopeStack::Cleanup {
+  cir::GlobalOp guard;
+  CallGuardAbort(cir::GlobalOp guard) : guard{guard} {}
+
+  void Emit(clang::CIRGen::CIRGenFunction &cgf, Flags flags) override {
+    auto guardPtrTy = cir::PointerType::get(guard.getSymType());
+    mlir::Value getGuard =
+        cgf.CGM.getBuilder().createGetGlobal(guard, /*threadLocal*/ false);
+    cgf.emitNounwindRuntimeCall(guard->getLoc(),
+                                getGuardAbortFn(cgf.CGM, guardPtrTy), getGuard);
+  }
+};
+} // namespace
+
+void LoweringPreparePass::handleStaticLocal(GlobalOp globalOp,
+                                            GetGlobalOp getGlobalOp) {
+
+  std::optional<cir::ASTVarDeclInterface> astOption = globalOp.getAst();
+  assert(astOption.has_value());
+  cir::ASTVarDeclInterface varDecl = astOption.value();
+
+  clang::CIRGen::CIRGenModule &cgm = *this->cgm;
+  clang::CIRGen::CIRGenCXXABI &cxxABI = cgm.getCXXABI();
+  clang::CIRGen::CIRGenBuilderTy &builder = cgm.getBuilder();
+  builder.setInsertionPointAfter(getGlobalOp);
+  Block *getGlobalOpBlock = builder.getInsertionBlock();
+  Operation *ret = getGlobalOpBlock->getTerminator();
+  ret->remove();
```

Comment:
This is hackery. But I might leave it as just a `TODO(CIR):` to expand the test case beyond `static local var a = foo(); return` and do it in another patch. This patch is big enough already.